### PR TITLE
Throw when meet empty PK

### LINKED ISSUE #6391: Throw exception when found empty pk
## Enhancement

Guard for https://github.com/pingcap/tiflash/issues/6379

### PR DESCRIPTION
**Pull request by CalvinNeo**
Signed-off-by: CalvinNeo <calvinneo1995@gmail.com>

### What problem does this PR solve?

Issue Number: close #6391

Problem Summary: throw exception earlier when meet illegal tikv key-value input like https://github.com/tikv/tikv/issues/13861

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

```
[2022/11/30 15:30:46.055 +08:00] [ERROR] [Exception.cpp:89] ["Code: 9008, e.displayText() = DB::Exception: Observe empty PK: raw key 7480000000000001FFBF5F720000000000FAF9ECEFDC3207FFFC, e.what() = DB::Exception, Stack trace:\n\n\n       0x36c272e\tStackTrace::StackTrace() [tiflash+57419566]\n

[2022/11/30 15:34:13.906 +08:00] [ERROR] [Exception.cpp:89] ["Code: 49, e.displayText() = DB::Exception: cur_offset does not equal to offset. is_common_handle true start_key  cur_offset 1 cur_limit 0 rows 1 offset 0 rowkey_range [01,03800000000000001D0380000000000000010380000000000000BF), e.what() = DB::Exception,
```

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/Transaction/RegionData.cpp [hunk 1]
```diff
@@ -115,7 +115,11 @@ RegionDataReadInfo RegionData::readDataByWriteIt(const ConstWriteCFIter & write_
     const auto & [pk, ts] = write_it->first;
 
     std::ignore = value;
-    std::ignore = value;
+
+    if (pk.toDebugString().empty())
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Transaction/RegionData.cpp:119]:
```suggestion
    if (pk->empty())
```

**Review by hehechen — APPROVED** [review]

File: dbms/src/Storages/Transaction/tests/gtest_new_kvstore.cpp [hunk 2]
```diff
@@ -154,6 +154,43 @@ TEST_F(RegionKVStoreTest, KVStoreFailRecovery)
...
+            MockRaftStoreProxy::FailCond cond;
+
+            auto kvr1 = kvs.getRegion(region_id);
+            auto r1 = proxy_instance->getRegion(region_id);
+            ASSERT_NE(r1, nullptr);
+            ASSERT_NE(kvr1, nullptr);
+            ASSERT_EQ(r1->getLatestAppliedIndex(), kvr1->appliedIndex());
+            {
+                r1->getLatestAppliedIndex();
+                std::string k = "7480000000000001FFBD5F720000000000FAF9ECEFDC3207FFFC";
```
Comment (lidezhu) [inline dbms/src/Storages/Transaction/tests/gtest_new_kvstore.cpp:180]: can you add some comment about the meaning of the `k` for easier later maintenance.

File: dbms/src/Debug/MockRaftStoreProxy.cpp [hunk 3]
```diff
@@ -381,6 +381,37 @@ std::tuple<uint64_t, uint64_t> MockRaftStoreProxy::normalWrite(
...
+        region->updateCommitIndex(index);
+        // We record them, as persisted raft log, for potential recovery.
+        region->commands[index] = {
+            term,
+            MockProxyRegion::RawWrite{
+                keys,
+                vals,
+                cmd_types,
+                cmd_cf,
+            }};
```
Comment (JaySon-Huang) [inline dbms/src/Debug/MockRaftStoreProxy.cpp:409]: I think `RawWrite` is more general than `NormalWrite`, maybe we can move the line `auto key = RecordKVFormat::genKey(table_id, keys[i], 1);` into `MockRaftStoreProxy::normalWrite` and get these two branches  merged into one 
https://github.com/pingcap/tiflash/blob/5b3a8b9a0346c030e47b9309345604f7cd094e2f/dbms/src/Debug/MockRaftStoreProxy.cpp#L461-L518

File: dbms/src/Storages/Transaction/tests/gtest_new_kvstore.cpp (hunk 2 above)
Comment (CalvinNeo) [reply dbms/src/Storages/Transaction/tests/gtest_new_kvstore.cpp:180]: This is just an invalid key(maybe truncated) TiKV sent...

File: dbms/src/Storages/Transaction/tests/gtest_new_kvstore.cpp [hunk 4]
```diff
@@ -154,6 +154,43 @@ TEST_F(RegionKVStoreTest, KVStoreFailRecovery)
     }
 }
 
+TEST_F(RegionKVStoreTest, KVStoreInvalidWrites)
+{
+    auto ctx = TiFlashTestEnv::getGlobalContext();
+    {
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Transaction/tests/gtest_new_kvstore.cpp:160]: we can remove this useless indent bracket

File: dbms/src/Common/RedactHelpers.cpp [hunk 5]
```diff
@@ -76,3 +76,17 @@ void Redact::keyToDebugString(const char * key, const size_t size, std::ostream
     }
     oss.flags(flags); // restore flags
 }
+
+std::string Redact::debugStringToKey(const char * start, size_t len)
```
[the remark this answers is not in the document]
Comment (CalvinNeo) [reply dbms/src/Common/RedactHelpers.cpp:80]: It just correspond to the previous `keyToDebugString`

[the remark this answers is not in the document]
Comment (lidezhu) [reply dbms/src/Common/RedactHelpers.cpp:80]: It seems that it is the reverse of `std::string Redact::keyToHexString(const char * key, size_t size)`. debugString may contains `?` in it.

File: dbms/src/Debug/MockRaftStoreProxy.cpp (hunk 3 above)
Comment (CalvinNeo) [reply dbms/src/Debug/MockRaftStoreProxy.cpp]: fixed

**Review by JaySon-Huang — APPROVED** [review]

File: dbms/src/Common/RedactHelpers.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (CalvinNeo) [reply dbms/src/Common/RedactHelpers.cpp]: fixed

**Pull request merged**